File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic.Syntax/ElseBlockSyntax.cs
using Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax;

namespace Microsoft.CodeAnalysis.VisualBasic.Syntax
{
	public sealed class ElseBlockSyntax : VisualBasicSyntaxNode
	{
		internal ElseStatementSyntax _elseStatement;

		internal SyntaxNode _statements;

		public ElseStatementSyntax ElseStatement => GetRedAtZero(ref _elseStatement);

		public SyntaxList<StatementSyntax> Statements
		{
			get
			{
				SyntaxNode red = GetRed(ref _statements, 1);
				return new SyntaxList<StatementSyntax>(red);
			}
		}

		internal ElseBlockSyntax(GreenNode green, SyntaxNode parent, int startLocation)
			: base(green, parent, startLocation)
		{
		}

		internal ElseBlockSyntax(SyntaxKind kind, DiagnosticInfo[] errors, SyntaxAnnotation[] annotations, ElseStatementSyntax elseStatement, SyntaxNode statements)
			: this(new Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ElseBlockSyntax(kind, errors, annotations, (Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ElseStatementSyntax)elseStatement.Green, statements?.Green), null, 0)
		{
		}

		public ElseBlockSyntax WithElseStatement(ElseStatementSyntax elseStatement)
		{
			return Update(elseStatement, Statements);
		}

		public ElseBlockSyntax WithStatements(SyntaxList<StatementSyntax> statements)
		{
			return Update(ElseStatement, statements);
		}

		public ElseBlockSyntax AddStatements(params StatementSyntax[] items)
		{
			return WithStatements(Statements.AddRange(items));
		}

		internal override SyntaxNode GetCachedSlot(int i)
		{
			return i switch
			{
				0 => _elseStatement, 
				1 => _statements, 
				_ => null, 
			};
		}

		internal override SyntaxNode GetNodeSlot(int i)
		{
			return i switch
			{
				0 => ElseStatement, 
				1 => GetRed(ref _statements, 1), 
				_ => null, 
			};
		}

		public override TResult Accept<TResult>(VisualBasicSyntaxVisitor<TResult> visitor)
		{
			return visitor.VisitElseBlock(this);
		}

		public override void Accept(VisualBasicSyntaxVisitor visitor)
		{
			visitor.VisitElseBlock(this);
		}

		public ElseBlockSyntax Update(ElseStatementSyntax elseStatement, SyntaxList<StatementSyntax> statements)
		{
			if (elseStatement != ElseStatement || statements != Statements)
			{
				ElseBlockSyntax elseBlockSyntax = SyntaxFactory.ElseBlock(elseStatement, statements);
				SyntaxAnnotation[] annotations = GetAnnotations();
				if (annotations != null && annotations.Length > 0)
				{
					return SyntaxNodeExtensions.WithAnnotations(elseBlockSyntax, annotations);
				}
				return elseBlockSyntax;
			}
			return this;
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/Syntax/ElseBlockSyntax.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
Imports System

Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax
	Public NotInheritable Class ElseBlockSyntax
		Inherits Microsoft.CodeAnalysis.VisualBasic.VisualBasicSyntaxNode
		Friend _elseStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.ElseStatementSyntax

		Friend _statements As SyntaxNode

		Public ReadOnly Property ElseStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.ElseStatementSyntax
			Get
				Return MyBase.GetRedAtZero(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.ElseStatementSyntax)(Me._elseStatement)
			End Get
		End Property

		Public ReadOnly Property Statements As SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.StatementSyntax)
			Get
				Return New SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.StatementSyntax)(MyBase.GetRed(Me._statements, 1))
			End Get
		End Property

		Friend Sub New(ByVal green As GreenNode, ByVal parent As SyntaxNode, ByVal startLocation As Integer)
			MyBase.New(green, parent, startLocation)
		End Sub

		Friend Sub New(ByVal kind As Microsoft.CodeAnalysis.VisualBasic.SyntaxKind, ByVal errors As DiagnosticInfo(), ByVal annotations As SyntaxAnnotation(), ByVal elseStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.ElseStatementSyntax, ByVal statements As SyntaxNode)
			MyClass.New(New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ElseBlockSyntax(kind, errors, annotations, DirectCast(elseStatement.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ElseStatementSyntax), If(statements IsNot Nothing, statements.Green, Nothing)), Nothing, 0)
		End Sub

		Public Overrides Function Accept(Of TResult)(ByVal visitor As VisualBasicSyntaxVisitor(Of TResult)) As TResult
			Return visitor.VisitElseBlock(Me)
		End Function

		Public Overrides Sub Accept(ByVal visitor As Microsoft.CodeAnalysis.VisualBasic.VisualBasicSyntaxVisitor)
			visitor.VisitElseBlock(Me)
		End Sub

		Public Function AddStatements(ByVal ParamArray items As Microsoft.CodeAnalysis.VisualBasic.Syntax.StatementSyntax()) As Microsoft.CodeAnalysis.VisualBasic.Syntax.ElseBlockSyntax
			Return Me.WithStatements(Me.Statements.AddRange(items))
		End Function

		Friend Overrides Function GetCachedSlot(ByVal i As Integer) As Microsoft.CodeAnalysis.SyntaxNode
			Dim syntaxNode As Microsoft.CodeAnalysis.SyntaxNode
			Dim num As Integer = i
			If (num = 0) Then
				syntaxNode = Me._elseStatement
			ElseIf (num = 1) Then
				syntaxNode = Me._statements
			Else
				syntaxNode = Nothing
			End If
			Return syntaxNode
		End Function

		Friend Overrides Function GetNodeSlot(ByVal i As Integer) As SyntaxNode
			Dim elseStatement As SyntaxNode
			Dim num As Integer = i
			If (num = 0) Then
				elseStatement = Me.ElseStatement
			ElseIf (num = 1) Then
				elseStatement = MyBase.GetRed(Me._statements, 1)
			Else
				elseStatement = Nothing
			End If
			Return elseStatement
		End Function

		Public Function Update(ByVal elseStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.ElseStatementSyntax, ByVal statements As SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.StatementSyntax)) As Microsoft.CodeAnalysis.VisualBasic.Syntax.ElseBlockSyntax
			Dim elseBlockSyntax As Microsoft.CodeAnalysis.VisualBasic.Syntax.ElseBlockSyntax
			If (elseStatement <> Me.ElseStatement OrElse statements <> Me.Statements) Then
				Dim elseBlockSyntax1 As Microsoft.CodeAnalysis.VisualBasic.Syntax.ElseBlockSyntax = Microsoft.CodeAnalysis.VisualBasic.SyntaxFactory.ElseBlock(elseStatement, statements)
				Dim annotations As SyntaxAnnotation() = MyBase.GetAnnotations()
				elseBlockSyntax = If(annotations Is Nothing OrElse CInt(annotations.Length) <= 0, elseBlockSyntax1, elseBlockSyntax1.WithAnnotations(annotations))
			Else
				elseBlockSyntax = Me
			End If
			Return elseBlockSyntax
		End Function

		Public Function WithElseStatement(ByVal elseStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.ElseStatementSyntax) As Microsoft.CodeAnalysis.VisualBasic.Syntax.ElseBlockSyntax
			Return Me.Update(elseStatement, Me.Statements)
		End Function

		Public Function WithStatements(ByVal statements As SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.StatementSyntax)) As Microsoft.CodeAnalysis.VisualBasic.Syntax.ElseBlockSyntax
			Return Me.Update(Me.ElseStatement, statements)
		End Function
	End Class
End Namespace